### ISSUE TITLE
More expansion errors in rewriter

### ISSUE THREAD

**Issue by wence-** [coneoproject/COFFEE#19]
firedrakeproject/firedrake#430 adds more tests to firedrake exercising COFFEE's rewriter when dealing with forms involving combinations of mathematical operators.  Many of them currently fail.  I guess it's probably all the same root cause, but I don't see it.

**Comment by FabioLuporini** [comment]
Could you write a minimal failing example? At least one that triggers one such error

**Comment by wence-** [comment]
So they're all minimally failing examples (I don't know if they all fail in the same way), but here's one:

```
from firedrake import *
mesh = UnitSquareMesh(1, 1)
V = FunctionSpace(mesh, 'CG', 1)
f = Function(V)
f.assign(1)
C = tanh(f)

u = TrialFunction(V)
v = TestFunction(V)

a = C*u*v*dx
L = C*v*dx

out = Function(V)

solve(a == L, out)

assert norm(assemble(out -1)) < 1e-6
```

Fails when on the ffc fix-funcall-gen branch.

**Comment by FabioLuporini** [comment]
Fix at the moment incorporated in branch fix-issues-8-18-19. There's a number of fixes in this branch, so there's a chance the other test cases can now pass. Let me now (but won't be able to work on this, problably, the next week, when I'll be in Amsterdam)

**Comment by wence-** [comment]
Yes, these now pass.

**Comment by FabioLuporini** [comment]
Closing since fixed by PR #21
